Remove Transactions::reload() and add `created_time` and `modified_time` to record and transaction models

bfish713 opened the pull request:
The transactions model would reload the in memory indexes from storage after every update to a transaction.  Since the index was changed to the persistable type with rollback, when we rebuild the index we enqueue an add change for every K,V pair in the index.  This means if we call revert we would end up removing every item from the index.  Obviously this is not intended.  This change fixes that by explicitly adding and removing only what changed in the index.  We only completely rebuild the index during create and immediately commit it so if the first storage transaction after creation fails we don't wipe the index.

Also this adds timestamps to `Record` and `Transaction` in the same pattern as the other models

jbearer commented on src/transactions.rs:
```diff
@@ -286,7 +291,13 @@ impl<L: Ledger> Transactions<L> {
         self.store.store(uid, txn)?;
-        self.reload();
+        if let Some(timeout) = txn.timeout() {
+            self.expiring_txns.insert((timeout, txn.uid().clone()));
```
If `txn.timeout()` was already set, this still adds an `Add` action to the changelog, causing the timeout to be removed from the index on a call to `revert`.

jbearer reviewed (changes requested):
This revert stuff is turning out to be really subtle.

Another thing I just realized, if a key already exists in a `PersistableHashMap`, then `insert` causes the value to be updated, but `revert` deletes the key entirely. I think we need to specialize our persistable data structures so that, e.g. `PersistableHashMap` is aware of the semantics of a `HashMap`,  likewise for `PersistableHashSet` (or, `PersistableHashSet<K> = PersistableHashMap<K, ()>`). And probably add an `Update` variant to `IndexChange`.

bfish713 commented:
I should have put this comment on here earlier, but when I put up the draft I realized there were a few issues with this still left to solve.  The thing you mentioned is an issue for sure.  Also if an update removes a value from `pending_uids` this will not be caught by the update logic either.  Revert would work fine since we simply don't remove the value from the index.  So the editor either needs to have access to `pending_uids` or there needs to be more logic which compares the entire set of uids for that transaction with the set in the index.   The problem is the index is mapping UIDs to transactions so it's hard to do from `Transactions`

shenkeyao commented:
The changes to add timestamps look good to me, but looks like the issue regarding the revert logic isn't fully resolved yet. Will there be a follow-up PR, or is this PR still in progress? I'll review again if it's the latter case.

bfish713 commented:
I think I resolved both issues, I've added handling for updates (inserts to HashMap where the key already exists) and fixed the logic for set types.  For sets if a an insert doesn't actually add a new item we should not enqueue a change in the pending_changes list.

For the issues with `uids_awaiting_memos` I was a bit mistaken because we don't ever update a uid in that index only add or remove.  There was still an issue with with removal where we would never revert a removal.  To fix this and the issue with inserting to this map I pushed the index updating logic to the editor layer

bfish713 commented:
> The changes to add timestamps look good to me, but looks like the issue regarding the revert logic isn't fully resolved yet. Will there be a follow-up PR, or is this PR still in progress? I'll review again if it's the latter case.

Oh I added you before I pushed my change and added the comment about what I did.  It should be good to review now with the revert logic changes

shenkeyao commented:
> The changes to add timestamps look good to me, but looks like the issue regarding the revert logic isn't fully resolved yet. Will there be a follow-up PR, or is this PR still in progress? I'll review again if it's the latter case.

Oh, for some reason the latest commits weren't displayed when I reviewed the PR, but now I can see them. I'll take a look.

jbearer commented on src/key_value_store.rs:
```diff
@@ -206,13 +213,18 @@ impl<K: Clone + Eq + Hash, V: Clone> Persist<(K, V)> for PersistableHashMap<K, V
-        self.index.remove(&change.0);
-        self.pending_changes.push(IndexChange::Remove(change));
+        if self.index.remove(&change.0).is_some() {
+            self.pending_changes.push(IndexChange::Remove(change));
```
Should be `Remove((change.0, old))`, I think. So that when we revert it, we are sure to add back in the old value.

bfish713 replied:
I guess I made the assumption that we were always passing in the actual value being removed but no reason to assume that is the case.

jbearer replied:
If that's true then fine, but we should probably assert that or something.

shenkeyao replied:
Maybe we should use the key rather than the `change` tuple as the input, and always get the old value from the index.

shenkeyao replied:
Oh but this will complicate the type of `change`, since for `PersistableHashSet` it's just `K` but for other structures it's a tuple. I guess it's good as is.

jbearer commented:
Another thing to think about (maybe long term) is we badly need test coverage for this reverting logic, since most of it is currently only triggered by exceptional error conditions that are test suite probably doesn't hit. This could take the form of error injection or more targeted unit tests that specifically test the `Persistable` interface, or both. For example, it would probably be pretty easy to write a comprehensive proptest that randomly generates sequences of operations, commits, and reverts on a persistable data structure and checks that it works as expected.
